contracts/command-batcher/src/encoding/evm.rs

```diff
@@ -0,0 +1,554 @@
+                message.source_chain,
+                message.source_address,
+                message.destination_address,
+                message.payload_hash,
+            ),
+            id: command_id(message.id),
+        }
+    }
+}
+
+#[cw_serde]
+pub struct Data {
+    pub destination_chain_id: Uint256,
+    pub commands: Vec<Command>,
+}
+
+impl Data {
+    fn new(destination_chain_id: Uint256, messages: Vec<Message>) -> Self {
+        let commands = messages.into_iter().map(|msg| msg.into()).collect();
+
+        Data {
+            destination_chain_id,
+            commands,
+        }
+    }
+
+    fn encode(&self) -> HexBinary {
+        let destination_chain_id = Token::Uint(
+            ethereum_types::U256::from_dec_str(&self.destination_chain_id.to_string())
+                .expect("violated invariant: Uint256 is not a valid EVM uint256"),
+        );
+
+        let mut commands_ids: Vec<Token> = Vec::new();
+        let mut commands_types: Vec<Token> = Vec::new();
+        let mut commands_params: Vec<Token> = Vec::new();
+
+        self.commands.iter().for_each(|command| {
+            commands_ids.push(Token::FixedBytes(command.id.to_vec()));
+            commands_types.push(Token::String(command.command_type.clone()));
+            commands_params.push(Token::Bytes(command.command_params.to_vec()));
+        });
+
+        ethabi::encode(&[
+            destination_chain_id,
+            Token::Array(commands_ids),
+            Token::Array(commands_types),
+            Token::Array(commands_params),
+        ])
+        .into()
+    }
+}
+
+impl Proof {
+    fn encode(&self) -> HexBinary {
+        let mut addresses: Vec<Token> = Vec::new();
+        let mut weights: Vec<Token> = Vec::new();
+        let mut signatures: Vec<Token> = Vec::new();
+
+        self.operators.iter().for_each(|operator| {
+            addresses.push(Token::Address(ethereum_types::Address::from_slice(
+                operator.address.as_slice(),
+            )));
+            weights.push(Token::Uint(
+                ethereum_types::U256::from_dec_str(&operator.weight.to_string())
+                    .expect("violated invariant: Uint256 is not a valid EVM uint256"),
+            ));
+
+            if let Some(signature) = &operator.signature {
+                signatures.push(Token::Bytes(signature.into()));
+            }
+        });
+
+        let threshold = Token::Uint(
+            ethereum_types::U256::from_dec_str(&self.threshold.to_string())
+                .expect("violated invariant: Uint256 is not a valid EVM uint256"),
+        );
+
+        ethabi::encode(&[
+            Token::Array(addresses),
+            Token::Array(weights),
+            threshold,
+            Token::Array(signatures),
+        ])
+        .into()
+    }
+}
+
+fn evm_address(pub_key: &[u8]) -> HexBinary {
+    let pub_key = PublicKey::from_sec1_bytes(pub_key)
+        .expect("violated invariant: pub_key is not a valid secp256k1 public key");
```

Comment:
I wasn't sure if this should be an error or a panic, my reasoning was that if it gets to the point of executing this code (keygen, key rotation and signing already happened) and this fails, then it's most likely a bug in code, because the actual validation should have happened first time the public key was introduced (I need to add this validation on the next PR where it includes keygen for the batcher contract)